### PR TITLE
Verb list page - select/deselect all verbs

### PR DESCRIPTION
**Pull request by JTrillo**
Pull request for issue [11](https://github.com/Lemoncode/english-quiz/issues/11), Verb list page - Allow deselecting all verbs, Allow selecting all verbs.
If you have any doubts or you think I should change anything, please tell me.

### CODE REVIEW

File: src/pods/configure-verbs/configure-verbs.component.tsx [hunk 1]
```diff
@@ -19,9 +20,11 @@ export const ConfigureVerbsComponent: React.FC<Props> = props => {
   const [temporalSelection, setTemporalSelection] = React.useState(
     verbCollection
   );
+  const [temporalSelected, setTemporalSelected] = React.useState(0);
```
Comment (brauliodiez) [inline src/pods/configure-verbs/configure-verbs.component.tsx:23]: Why not enclosing this plus business into a hook,

File: src/pods/configure-verbs/configure-verbs.component.tsx [hunk 2]
```diff
@@ -31,13 +34,30 @@ export const ConfigureVerbsComponent: React.FC<Props> = props => {
       const index = draft.findIndex(item => item.verbKey === verbId);
 
       if (index !== -1) {
+        if (draft[index].selected) {
+          setTemporalSelected(temporalSelected-1);
+        } else {
+          setTemporalSelected(temporalSelected+1);
+        }
```
Comment (brauliodiez) [inline src/pods/configure-verbs/configure-verbs.component.tsx:41]: I think we could enclose all this complexity in the hook

File: src/pods/configure-verbs/configure-verbs.component.tsx [hunk 3]
```diff
@@ -31,13 +34,30 @@ export const ConfigureVerbsComponent: React.FC<Props> = props => {
...
+  const handleMasterCheckboxChange = () => (
+    e: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    if (temporalSelected !== verbCollection.length) {
+      setTemporalSelection(selectOrDeselectAll(temporalSelection, true));
+      setTemporalSelected(verbCollection.length)
+    } else {
+      setTemporalSelection(selectOrDeselectAll(temporalSelection, false));
+      setTemporalSelected(0);
+    }
```
Comment (brauliodiez) [inline src/pods/configure-verbs/configure-verbs.component.tsx:58]: Same here candidate to be enclosed in the selection

File: src/pods/configure-verbs/configure-verbs.component.tsx [hunk 4]
```diff
@@ -53,6 +73,22 @@ export const ConfigureVerbsComponent: React.FC<Props> = props => {
...
+          <FormControlLabel
+            control={
+              <Checkbox
+                checked={temporalSelected !== 0}
+                onChange={handleMasterCheckboxChange()}
+                color="primary"
+                indeterminate={
+                  temporalSelected !== 0 &&
+                  temporalSelected !== verbCollection.length
+                }
```
Comment (brauliodiez) [inline src/pods/configure-verbs/configure-verbs.component.tsx:86]: Same thing this we can enclose this logic check inside the hooks and provide some meaningful name, e.g.

indetermainate{SelectionMatchesItemsChecked()}

**Comment by JTrillo** [comment]
Done.

File: src/pods/configure-verbs/configure-verbs.component.tsx [hunk 5]
```diff
@@ -15,30 +15,68 @@ interface Props {
...
 }
 
-export const ConfigureVerbsComponent: React.FC<Props> = props => {
-  const { verbCollection, onSave, onCancel } = props;
-  const [temporalSelection, setTemporalSelection] = React.useState(
-    verbCollection
-  );
+const useTemporalSelectionAndSelected = (verbCollection: VerbEntity[]) => {
+  const [temporalSelection, setTemporalSelection] = React.useState<VerbEntity[]>([]);
   const [temporalSelected, setTemporalSelected] = React.useState(0);
```
Comment (brauliodiez) [inline src/pods/configure-verbs/configure-verbs.component.tsx:20]: By temporalSelected, you mean how many items are selected? Is that equales to temporalSelection.length?

Could you please ellaborate a bit more? Thanks

File: src/pods/configure-verbs/configure-verbs.component.tsx [hunk 6]
```diff
@@ -15,30 +15,68 @@ interface Props {
...
     setTemporalSelected(getSelectedNumber(verbCollection));
   }, [verbCollection]);
 
+  const increaseOrDecreaseTemporalSelected = (selected: boolean) => {
+    if (selected) {
+      setTemporalSelected(temporalSelected - 1);
+    } else {
+      setTemporalSelected(temporalSelected + 1);
+    }
+  };
```
Comment (brauliodiez) [inline src/pods/configure-verbs/configure-verbs.component.tsx:33]: Maybe we can split it in two methods

increaseTemporalSelected

decreaseTemporalSelected

File: src/pods/configure-verbs/configure-verbs.component.tsx (hunk 5 above)
Comment (brauliodiez) [reply src/pods/configure-verbs/configure-verbs.component.tsx:20]: I see, the list can contain a different number of items selected, I need to understand a bit deeper how this logic works, could you please ellaborate?

File: src/pods/configure-verbs/configure-verbs.component.tsx [hunk 7]
```diff
@@ -15,30 +15,68 @@ interface Props {
...
+
+  const updateAfterMasterCheckboxChange = () => {
```
Comment (brauliodiez) [inline src/pods/configure-verbs/configure-verbs.component.tsx:35]: ResetSelectionBasedOnRootSelectionCheckBox ?

File: src/pods/configure-verbs/configure-verbs.component.tsx (hunk 5 above)
Comment (JTrillo) [reply src/pods/configure-verbs/configure-verbs.component.tsx:20]: Ok, I explain it.
_**temporalSelection**_ has always the same quantity of elements, only the field _**selected**_ on the elements changes.
So I needed another state to store how many verbs are selected at the moment. In a first approach, I tried to calculate this number using function **getSelectedNumber** (_configure-verbs.business.ts_) without having the state. But this is inefficient, it is better to call **getSelectedNumber**  once, when the component is mounted.
Why do I need the currect number of select verbs? For the master checkbox (when _**temporalSelection**_ = 0, is unchecked; when >= 0 & <= total, is indeterminate; when = total, is checked)

File: src/pods/configure-verbs/configure-verbs.component.tsx (hunk 6 above)
Comment (JTrillo) [reply src/pods/configure-verbs/configure-verbs.component.tsx:33]: I thought the same, but I declared one finally in order to encapsule conditionals in it. If you prefer, I can split **increaseOrDecreaseTemporalSelected** in two methods as you say.

File: src/pods/configure-verbs/configure-verbs.component.tsx (hunk 7 above)
Comment (JTrillo) [reply src/pods/configure-verbs/configure-verbs.component.tsx:35]: Ok, I will change the name of the method.

File: src/pods/configure-verbs/configure-verbs.component.tsx (hunk 5 above)
Comment (brauliodiez) [reply src/pods/configure-verbs/configure-verbs.component.tsx:20]: Mmmm... we could take advantage of the hooks and the ability to execute some code just whenever a variable changes, something like

```tsx
// calculateRootCheckBoxState
React.useEffect(() => {
  // Do your calculation thing, is zero? is equals to length? ...
  // set the checkbox status
}, [temporalSelection,]);
```
By doing this you only execute this code whenever temporalSelection gets updated and you save having temporary variables.

What do you think?

**Comment by JTrillo** [comment]
Done.

**Comment by JTrillo** [comment]
Done.

**Pull request merged**